### ISSUE TITLE
Refactoring tests.

### ISSUE THREAD

**Issue by ndobromirov** [lcache/lcache#43]
The current tests cover everything in 1 file. This include pure unit tests for all classes and integration tests for checking whether all of them actually work together. This is already close to 1000 lines of code.

It would be nice to have this separated, following the single responsibility OO principle. So there should be:
1. PHPUnit test class to test every data object and interface (Entry, Address, others if i've missed any).
2. Utility classes for testing L1 in isolation against all the currently available L2 implementations.
3. Utility classes for testing L2 in isolation against all the currently available L1 implementations.
4. Integration test base (or a trait) that will test interoperability of a given L1 + L2 interoperability to reduce code duplication in 2 and/or 3.

This way adding a new L1 for example would be to create L1 implementation class and a test class, following TDD and improving maintainability.

**Comment by davidstrauss** [comment]
Yes, this is definitely something I'd like to do. I've been meaning to look into best practices for automatically testing every permutation of L1 and L2, and this advice will get me there quicker.

**Comment by ndobromirov** [comment]
This overlaps with the work done on #72 . 
Decide whether to close this one and extend the other issue a bit.

**Comment by stevector** [comment]
@ndobromirov, sorry I should have reviewed the issue queue before opening https://github.com/lcache/lcache/issues/70 😄  I think we're looking for the same kind of refactor. 

I'll close https://github.com/lcache/lcache/issues/70.

https://github.com/lcache/lcache/issues/72 is not ready to be merged as it stands now. I need to step it back a bit so that it only refactors the L1 pieces. I'm using traits there as a helper. But maybe you had a different implementation in mind for code reuse? Can you give me an example of what you meant by 
>Utility classes for testing L1 in isolation against all the currently available L2 implementations.

**Comment by davidstrauss** [comment]
Just so folks know, I'm not trying to be AWOL here; I'm just hacking on a RocksDB extension for PHP to try as an ideal L1 backend.

**Comment by ndobromirov** [comment]
@stevector, the general idea was to have:
1. L1 tests for each driver - almost done in your PR with traits. My idea was to have an abstract L1 test base class with 1 abstract method - factory for the current L1 instance being tested and all the actual tests to be implemented there. Having all the L1 concrete tests implemented by sub-classing that one seems optimal, as they should confirm to L1 interface and inherit all the test methods from there and they should not differ functionally at tall, if they do (the test should be specific in the given concrete implementation), overwriting the test method or fixing the compatibility issue..
2. L2 tests for each driver (against all available L1s, where needed) - TBD (there is something in your PR also). This should follow the Idea as in L1, tough this tests should fully depend on all L1 tests passing, As testing L2 depends on Having a working L1. Any ideas are welcome here.. 
3. And then test the integration class with all combinations of L1 and L2 available in the project. So the quote from @stevector is related to having this last point. I might have not been very clear explaining it. Though the actual testing implementation is open for discussion. I have 2 ideas at the moment for this:
3.1. Have 2 utility methods that provide ready to use instances of L1 and L2. So we just have a single test method that generates all the possible combinations for integration instances to test and then have all the tests for that integration instance tested.
3.2. Follow the idea from L1 and L2 tests... To have an abstract base class that requires a factory. Then have a concrete implementation that will test a single combination of L1 and L2 for the integration layer. This is not my favorite, as there are going to be multitude of test classes (amount of files to maintain). That is going to be (count L1s) * (count L2s).
3.3. Any better ideas are welcome :).

@davidstrauss what is the extension you are testing with?
When I was researching RocksDB there were 3-4 that I found but all of them were in kind of POC state or legacy (for PHP 5.5).

**Comment by davidstrauss** [comment]
> My idea was to have an abstract L1 test base class with 1 abstract method - factory for the current L1 instance being tested and all the actual tests to be implemented there.

This is an excellent -- and very clever -- idea. I hadn't considered having the test class hierarchy follow the actual object hierarchy. It's also important to me that most tests run against most implementations of the same layer (L1, L2, etc.).

I also agree with the other testing ideas, too.

> what is the extension you are testing with?

I'm writing one. There's a lot to do, still:

1. Improve RocksDB to implement the necessary features. RocksDB already has a TTL-based wrapper, but it needs support for dynamic (at write time) expirations. After some prototype implementations, I've agreed with Facebook's engineers on a way to add this.
2. Expose necessary features through the C binding. Some of the necessary features area available only via C++ right now. Obviously, new features will also need such wrappers.
3. Write the PHP extension. I have a shell going already that links to RocksDB and performs some basic operations to prove the linking and configuration work. With the work in the prior items, a PHP wrapper that provides a nearly equivalent API to APCu (but on top of RocksDB) should be easy. RocksDB will handle cache size limitations, pruning expired items, and counters.

**Comment by ndobromirov** [comment]
Going through the code-base, of level 2 drivers there seems to be no concept of pooling. Please bare with me for a bit.

As I understand it, there is **pool** configurations that will separate different clients / applications on level 1 so here are no data conflicts. And then this cache pool is separated by the different **bins** - OK.

There is no such concept in L2, so by definition L1 and L2 to work in the same pool. It is kind of enforced by the integration layer. Though the pool data is present in a column and as part of the **Entry** / **event** objects. The thing is, it's a bit confusing to be passed as parameter on the set method, when it can be known on instantiation. :thinking: 

It sees that L2 is made to be pool-agnostic, so it will not require such meta-data upon creation, but this will open issues like calling L2 set against a different pool (just in theory possible with a custom integration class) causing data corruption in different application (if a collision is made).

I do not see why L2 instance will not have pool data passed on. It will be paired with a pooled L1 either way and they will be mapped 1:1 as best as they can. Am I missing something, causing my confusion or it is just confusing :D?

Some other off-topics:
 - The DatabaseL2 and StaticL2 are kind of non-compatible as interface - the db has a global auto-increment id (enforced by the DB). Different DatabaseL2 instances will share this ID and increment onwards. The static one has an instance-bound ID, so any new L2 instance will start from 0. At the moment there is no way to share L2 state in the static driver, unless you share the driver's instance (a pool value would have helped here). How should that work on static? I think they should have the same behaviour.
 - The DatabaseL2 has the concept of clean-up (prunning ) of stale non-needed events data. The static one does not, even if such might be needed in long-running processes (many tests). Or this is not needed, because L2 will never be used in static context and the static implementation is just for tests. But then it does not comply with the DatabaseL2 behaviour. :thinking: 

I will sleep on it... The morning is always smarter :).

**Comment by ndobromirov** [comment]
Please ignore the above comment, as apparently it was very late :).

Just to also ping here for the progress, I think it is getting very close with all the changes and tuning of L2. 
I will be starting the integration layer soon. Anything there will be based on the PR #79.

**Comment by davidstrauss** [comment]
> And then this cache pool is separated by the different bins - OK.

Pools and bins are orthogonal (not a hierarchy). Every cache item is written to a bin (often a default one for many clients, like Drupal's), but the pool only tracks which PHP-FPM pool originated the event. This is to avoid informing cache pools about events they already know about because they created them.

> The thing is, it's a bit confusing to be passed as parameter on the set method, when it can be known on instantiation.

It's mostly for testing so we can run different `Integrated` clients against the same `L2`. I'm not sure I can think of *production* reasons why you would ever want an instantiated `L2` to write with different pool IDs at different times. For things like the `StaticL2`, it would be difficult to properly share an L2 in testing if we had to fix the L2 to a single pool on instantiation (more on this below).

> It sees that L2 is made to be pool-agnostic, so it will not require such meta-data upon creation, but this will open issues like calling L2 set against a different pool (just in theory possible with a custom integration class) causing data corruption in different application (if a collision is made).

I'm not sure I would call it corruption as long as you're using a separate `L1`. LCache guarantees that any `Integrated` instance (which you would need to use a different `L1`) can see all cache writes preceding the call to `Integrated->synchronize()`. There's no guarantee that a second `Integrated` instance in the same request will have its writes propagate to or from the the other `Integrated` instance. Multiple tests rely on this property to test synchronization within one process of one test run.

Instantiating an `Integrated` instance and running `->synchronize()` is sort of like `BEGIN TRANSATION` in SQL with isolation set to `REPEATABLE READ`. Basically, you shouldn't expect to see changes that have occurred since you instantiated. LCache is weaker than `REPEATABLE READ`, though, as you *might* see some writes since sync.

Fortunately, in practice, web applications:

- Expect that cache values may change if they re-read them, even if they didn't alter them. This is very true for memcached, Redis, and other caches. It is alsol true for LCache. No problem here.
- Expect to get cache values at least as fresh as when the request started. This is important for LCache, as systems like memcached *will* show you the latest value whenever you ask, unlike LCache.
- Do not expect to read cache updates by other requests during a request. The only exception I can think of here is trying to run locks with polling, but that would be dangerous without relying on atomic operations, which LCache doesn't provide or promise. It is also really inefficient.

> The DatabaseL2 and StaticL2 are kind of non-compatible as interface - the db has a global auto-increment id (enforced by the DB). Different DatabaseL2 instances will share this ID and increment onwards. The static one has an instance-bound ID, so any new L2 instance will start from 0. At the moment there is no way to share L2 state in the static driver, unless you share the driver's instance (a pool value would have helped here). How should that work on static? I think they should have the same behaviour.

It's behaving as designed. The `StaticL2` driver is mostly a mock for testing. I would consider having it take in an array by reference for state, though, if we wanted to instantiate `L2` caches with a single pool. We could then move both events and the event counter to part of the array, allowing you to instantiate two `StaticL2`s in the same request that share state. I'm not sure how much benefit it would be, though, as `StaticL2` is mostly for testing `Integrated` and `L1`; isolating shared state would only help to test `StaticL2` better.

> The DatabaseL2 has the concept of clean-up (prunning ) of stale non-needed events data. The static one does not, even if such might be needed in long-running processes (many tests). Or this is not needed, because L2 will never be used in static context and the static implementation is just for tests. But then it does not comply with the DatabaseL2 behaviour.

Your "this is not needed" perspective is correct; `StaticL2` is for testing. That said, I have no problem with adding cleanup to it for better consistency with `DatabaseL2`. It would not save memory for long test runs, though, based on how PHP manages memory. We basically discard the entire `StaticL2` between tests, so its ability to prune away some state isn't necessary for performance or memory efficiency.

**Comment by ndobromirov** [comment]
> I'm not sure I would call it corruption as long as you're using a separate L1.

Exactly my notion of corruption was when you share servers  for different applications running in the same server + port in different folders under web-root. Then the default assumption of **SERVER_ADDR + SERVER_PORT** does not work. I think this is more of a documentation issue than library's code one - that users should check the pool value uniqueness if doing things like servers sharing. This is more of an issue for L1, as L2 is likely to be in a different database (for the different app), unless the DB is shared in some crazy way :D.

> It's behaving as designed.

In #79 there is a small tweak in the StaticL2 state storage - moved to a static class variable, so all instances of the class will share it, making it behave more like the DB version of L2. This can be reverted and tests tweaked to work with it.

> Your "this is not needed" perspective is correct; StaticL2 is for testing. That said, I have no problem with adding cleanup to it for better consistency with DatabaseL2. It would not save memory for long test runs, though, based on how PHP manages memory. We basically discard the entire StaticL2 between tests, so its ability to prune away some state isn't necessary for performance or memory efficiency.

Some tests were expecting clean-up that the static driver did not have at the time of writing. When I've implemented the tests to cover both Static and Database L2s all of them failed. This clean-up is implemented in #79.

**Comment by davidstrauss** [comment]
Yes, I'm keeping your change to make StaticL2 share data across instances.
